[Ready] Be more aggressive classifying Roundabout Intersections

daniel-j-h opened the pull request:
Roundabout Intersections are roundabouts with up to four ways and turn
angles which makes the turns obvious, e.g. as in:

```
    *
    *
* *   * *
    *
    *
```

but not e.g.

```
     *
    *
* *   *
    * *
     * *
```

For Roundabout Intersections we issue instructions such as
"turn \<direction\>" instead of "take the \<nth\> exit".

At the moment we have a limit on the radius for these Roundabout
Intersections of 5 meters. Which fails to classify a wide range of
Roundabout Intersections in the US (with the US-wide streets).

Example: http://www.openstreetmap.org/way/348812150

This changeset removes the Roundabout Intersection radius limit:

- if the roundabout is larger than a threshold and is named we classify
  it as a rotary

- if the roundabout matches our criteria for Roundabout Intersections
  we classify it as a Roundabout Intersection (note: we no longer care
  about its radius here)

- else fall back to plain old Roundabout

Up for discussion. cc @1ec5 @MoKob @TheMarex

There is a second issue with determining a roundabout's radius.
But that's for another pull request (tracking in #2716).

daniel-j-h commented on features/guidance/roundabout-turn.feature:
```diff
@@ -389,3 +389,44 @@ Feature: Basic Roundabout
            | a,h       | ab,gh,gh | depart,roundabout turn right exit-1,arrive    | 0->180,180->224,270->0 |
 
+
+    Scenario: Large radius Roundabout Intersection and ways modelled out: East Mission St, North 7th St
```
This scenario is new, the other ones are adapted to Roundabout Intersections.

1ec5 commented:
That sounds reasonable for the case given above, which is essentially just a traffic calming device. Do you recall how the five-meter threshold was determined? Are there counterexamples in Europe where a five-meter-wide roundabout would be a conventional junction rather than a traffic-calming device?

daniel-j-h commented:
I don't recall the 5 meters heuristic --- could be totally guestimated.
We have to wait for @MoKob for insight here.

My main point is: once we have determined it's not a rotary (large named roundabout), does the size actually matter for classifying these as roundabout intersections?

As long as the roundabout intersection criteria are met, I see no reason (other than maybe mis-tagged rotaries?) to not announce them as turns.

emiltin commented:
i agree that it makes sense to announce turns if it's not a rotary and the turns are almost 90 degree. for example this series of roundabouts:
http://www.openstreetmap.org/#map=16/55.6564/12.5864

in additional to turns being almost 90 degree, there clearly a main road continuing straight ahead, which makes "straight ahead" even more natural.

MoKob commented on features/guidance/roundabout-bike.feature:
```diff
@@ -22,19 +22,19 @@ Feature: Basic Roundabout
+           | h,a       | gh,ab,ab | depart,roundabout turn left exit-3,arrive     |
+           | h,d       | gh,cd,cd | depart,roundabout turn straight exit-2,arrive |
+           | h,f       | gh,ef,ef | depart,roundabout turn right exit-1,arrive    |
 
```
if we want to have it classify as a turn, this should not only be adjusted but moved to the intersection features.

MoKob commented on features/guidance/roundabout.feature:
```diff
@@ -22,19 +22,19 @@ Feature: Basic Roundabout
-           | h,a       | gh,ab,ab | depart,roundabout-exit-3,arrive |
-           | h,d       | gh,cd,cd | depart,roundabout-exit-2,arrive |
-           | h,f       | gh,ef,ef | depart,roundabout-exit-1,arrive |
+           | waypoints | route    | turns                                         |
```
if kept, this also needs to move location.

MoKob commented on features/guidance/roundabout.feature:
```diff
@@ -295,9 +295,9 @@ Feature: Basic Roundabout
-            | waypoints | route    | turns                           |
-            | a,e       | ab,ce,ce | depart,roundabout-exit-1,arrive |
-            | a,f       | ab,df,df | depart,roundabout-exit-2,arrive |
+            | waypoints | route    | turns                                         |
```
same as above

MoKob commented on src/extractor/guidance/roundabout_handler.cpp:
```diff
@@ -329,23 +326,13 @@ RoundaboutType RoundaboutHandler::getRoundaboutType(const NodeID nid) const
-            return RoundaboutType::Roundabout;
     }
 
-    if (radius <= MAX_ROUNDABOUT_INTERSECTION_RADIUS)
```
I find it highly dangerous not to have a radius check in here. There are roundabouts of more than 2km in range (special case, though real). Announcing all of them as turns is very dangerous.

Also, if we have a long distance traveled in the roundabout, it is likely that `straight` is not obvious anymore.

daniel-j-h replied:
Thanks for flagging this, I simply had no context why you added the limit there. Is this a data issue or a real intersection following the roundabout rules?

MoKob replied:
In this case it is actually a incredibly large loop that really follows roundabout rules. Not exactly sure anymore were it was, but a friend of mine working at a different navigation company told me about this case.

TheMarex replied:
In cases like this it would be helpful IMHO to add some context to the code. Link to OSM and a few words should suffices. Best would of course be a cucumber scenario testing for this behavior.

MoKob replied:
There were a few cucumber cases testing roundabout sizes. They have been adjusted for this change (see above).

daniel-j-h commented:
Adopted code to comments: added upper limit for roundabout intersections, moved the cucumber tests into the roundabout turns file, created roundabout turn bike test. Rebased against master. Waiting for Travis, then ready to go.
